peer.go

```diff
@@ -0,0 +1,1087 @@
+	controlCh := make(chan peerControl)
+	p := makePeerInternal(serverID, serverAddress, logs, snapshots,
+		goRoutines, trans, localAddr, controlCh, progressCh, options)
+	p.shared.goRoutines.spawn(p.selectLoop)
+	return controlCh
+}
+
+func makePeerInternal(serverID ServerID,
+	serverAddress ServerAddress,
+	logs LogStore,
+	snapshots SnapshotStore,
+	goRoutines *waitGroup,
+	trans Transport,
+	localAddr ServerAddress,
+	controlCh <-chan peerControl,
+	progressCh chan<- peerProgress,
+	options peerOptions) *peerState {
+
+	if options.logger == nil {
+		options.logger = log.New(os.Stderr, "", log.LstdFlags)
+	}
+	if options.maxAppendEntries == 0 {
+		options.maxAppendEntries = 1000
+	}
+	if options.heartbeatInterval == 0 {
+		options.heartbeatInterval = 100 * time.Millisecond
+	}
+	if options.failureWait == 0 {
+		options.failureWait = 10 * time.Millisecond
+	}
+	if options.maxFailureWait == 0 {
+		options.maxFailureWait = 100 * time.Millisecond
+	}
+
+	p := &peerState{
+		shared: &peerShared{
+			options:            options,
+			peerID:             serverID,
+			peerAddr:           serverAddress,
+			trans:              trans,
+			localAddr:          localAddr,
+			logger:             options.logger,
+			logs:               logs,
+			snapshots:          snapshots,
+			goRoutines:         goRoutines,
+			requestCh:          make(chan peerRPC),
+			replyCh:            make(chan peerRPC),
+			pipelineSendDoneCh: make(chan struct{}),
+			stopCh:             make(chan struct{}),
+		},
+		controlCh:  controlCh,
+		progressCh: progressCh,
+
+		backoffTimer: time.NewTimer(time.Hour),
+		progress: peerProgress{
+			peerID: serverID,
+		},
+	}
+	p.backoffTimer.Stop()
+
+	pipeline, err := p.shared.trans.AppendEntriesPipeline(p.shared.peerAddr)
+	if err != nil {
+		if err != ErrPipelineReplicationNotSupported {
+			p.shared.logger.Printf("[ERR] raft: Failed to start pipeline replication to %s: %s",
+				p.shared.peerAddr, err)
+		}
+	} else {
+		p.shared.pipeline = pipeline
+	}
+
+	return p
+}
+
+func (p *peerState) checkInvariants() error {
+	if p.shared.peerID != p.progress.peerID {
+		return fmt.Errorf("Progress must have peerID")
+	}
+	if p.progress.lastContact.After(p.progress.lastReply) ||
+		p.progress.lastContact == p.progress.lastReply && !p.progress.lastContact.IsZero() {
+		return fmt.Errorf("lastContact must be after lastReply (or both 0)")
+	}
+	if p.nextIndex > p.control.lastIndex+1 {
+		return fmt.Errorf("nextIndex must not be more than 1 entry past the lastIndex")
+	}
+	if p.outstandingInstallSnapshotRPC && p.outstandingAppendEntriesRPCs > 0 {
+		return fmt.Errorf("must not send a snapshot and entries simultaneously")
+	}
+	return nil
+}
+
+// selectLoop is a long-running routine that sends RPCs to a single remote
+// server.
+func (p *peerState) selectLoop() {
+	for {
+		// At this point, we may or may not have more work that we could do right away.
+		// Check for new information but don't block.
+		p.drainNonblocking()
```

Comment:
Are there correctness concerns with the drain or is this a performance optimization to avoid calling `issueWork()` too often? It seems like we could simplify things a bit by always doing a blocking select and then calling `issueWork()` each time. It seems like the one special case we might want would be something like this:

`
for {
  select {
    // blocking select here over everything
  }

  select {
    // do a non-blocking poll of just the control channel one time to make sure we have latest
  }

  p.issueWork()
}
```

That would ensure we are getting the latest control info before we respond to any other operation.